cylc/uiserver/authorise.py

```diff
@@ -0,0 +1,476 @@
+        """
+        Returns operation name required for authorization.
+        Converts queries and subscriptions to read operations.
+        Args:
+            field_name: Field name e.g. play
+            operation: operation type
+
+        Returns:
+            operation name
+        """
+        if operation in Authorization.READ_AUTH_OPS:
+            return Authorization.READ_OPERATION
+        else:
+            # Check it is a mutation in our schema
+            if (field_name in Authorization.ALL_OPS):
+                return field_name
+        return None
+
+    async def async_resolve(self, next_, root, info, **args):
+        """Return awaited coroutine"""
+        return await next_(root, info, **args)
+
+
+def get_groups(username: str) -> List[str]:
+    """Return list of system groups for given user
+
+    Args:
+        username: username used to check system groups.
+
+    Returns:
+        list: system groups for username given
+    """
+    groupmax = (os.NGROUPS_MAX)  # type: ignore
+    group_ids = os.getgrouplist(username, groupmax)
+    group_ids.remove(groupmax)
+    # turn list of group_ids into group names with group identifier prepended
+    return list(map(
+        lambda x: f'{Authorization.GRP_IDENTIFIER}{grp.getgrgid(x).gr_name}',
+        group_ids
+        )
+    )
+
+
+def expand_and_process_access_groups(permission_set: set) -> set:
+    """Process a permission set.
+
+    Takes a permission set, e.g. limits, defaults.
+    Expands the access groups and removes negated operations.
+
+    Args:
+        permission_set: set of permissions
+
+    Returns:
+        permission_set: processed permission set.
+    """
+    for action_group, expansion in {
+        Authorization.CONTROL: Authorization.CONTROL_OPS,
+        Authorization.ALL: Authorization.ALL_OPS,
+            Authorization.READ: Authorization.READ_OPS}.items():
+        if action_group in permission_set:
+            permission_set.remove(action_group)
+            permission_set.update(expansion)
+    # Expand negated permissions
+    for action_group, expansion in {
+            Authorization.NOT_CONTROL: list(
+                map((lambda x: '!' + x), Authorization.CONTROL_OPS)),
+            Authorization.NOT_ALL: list(
+                map((lambda x: '!' + x), Authorization.ALL_OPS)),
+            Authorization.NOT_READ: list(
+                map((lambda x: '!' + x), Authorization.READ_OPS))}.items():
+        if action_group in permission_set:
+            permission_set.remove(action_group)
+            permission_set.update(expansion)
+    # Remove negated permissions
+    remove = set()
+
+    for perm in permission_set:
+        if perm.startswith("!"):
+            with suppress(KeyError):
+                remove.add(perm.lstrip("!"))
+                remove.add(perm)
+    permission_set.difference_update(remove)
+    permission_set.discard('')
+    return permission_set
```

Comment:
Should we log if the user passes a permission that does not exist? e.g.:

```
"group:group1": {"default": "TRUCK" }
```

Maybe warn too - if possible - if a key is duplicated? (I don't think we can though...), e.g.:

```
"group:group1": {"default": "ALL", "default": "TRUCK" }
```

Reply:
(Tested and the UIS loaded fine)

Reply:
For the duplicated keys, currently the 2nd "default" will override the first. I think this is probably acceptable to document? 

I have left a `TODO` for more rigorous dictionary validating in `_check_site_auth_dict_correct_format`, we could possibly warn about duplicates and non-existent permissions here. I wanted to do more reading on validating traitlets before embarking on this one!

Passing in non-existent permissions is I guess not a big deal in terms of security (any operations executed by users are checked against our schema mutations, so nothing can slip through here) but it could be frustrating for users if, for example, a typo and they intend to give permissions which are not enacted.

This leads me back to thinking that it would be nice to have an interface whereby users could check what permissions a user has for their uiserver, so, put in a username and it generates a list of permissions for that user. I think this could be helpful due to the support for groups. However, I'm not sure how widely used this feature will be...possibly not worth the time investment!

Reply:
>This leads me back to thinking that it would be nice to have an interface whereby users could check what permissions a user has for their uiserver, so, put in a username and it generates a list of permissions for that user. I think this could be helpful due to the support for groups. However, I'm not sure how widely used this feature will be...possibly not worth the time investment!

This tool does sound useful. Maybe create an issue here or in cylc/cylc-ui with low priority?